src/lib/storage/dictionary_column.hpp

```diff
@@ -0,0 +1,116 @@
+#pragma once
+
+#include <algorithm>
+#include <limits>
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "all_type_variant.hpp"
+#include "fitted_attribute_vector.hpp"
+#include "type_cast.hpp"
+#include "types.hpp"
+#include "value_column.hpp"
+
+namespace opossum {
+
+class BaseAttributeVector;
+class BaseColumn;
+
+// Even though ValueIDs do not have to use the full width of ValueID (uint32_t), this will also work for smaller ValueID
+// types (uint8_t, uint16_t) since after a down-cast INVALID_VALUE_ID will look like their numeric_limit::max()
+constexpr ValueID INVALID_VALUE_ID{std::numeric_limits<ValueID::base_type>::max()};
+
+// Dictionary is a specific column type that stores all its values in a vector
+template <typename T>
+class DictionaryColumn : public BaseColumn {
+ public:
+  /**
+   * Creates a Dictionary column from a given value column.
+   */
+  explicit DictionaryColumn(const std::shared_ptr<BaseColumn>& base_column) { this->_compress(base_column); }
+
+  // return the value at a certain position. If you want to write efficient operators, back off!
+  const AllTypeVariant operator[](const size_t i) const override { return _dictionary->at(_attribute_vector->get(i)); }
+
+  // return the value at a certain position.
+  const T get(const size_t i) { return (*_dictionary)[_attribute_vector->get(i)]; }
```

Comment:
Why are you using `_dictionary->at(_attribute_vector->get(i))` in `operator[]` but `(*_dictionary)[_attribute_vector->get(i)]` here? Most notably, there is no bounds check here.